### ISSUE TITLE
The Jan. 23 5.8/trunk snapshots onwards crash when compiling NIO trunk for linux/Android

### ISSUE THREAD

**Issue by buttaface** [apple/swift#63262]
**Description**
I've started seeing this crash with the last few snapshots:
```
swift-frontend: /home/build-user/swift/lib/IDE/IDETypeChecking.cpp:945: SmallVector<std::pair<ValueDecl *, ValueDecl *>, 1> swift::getShorthandShadows(swift::CaptureListExpr *, swift::DeclContext *): Assertion `DeclaredVar->getName() == ReferencedVar->getName()' failed.
Please submit a bug report (https://swift.org/contributing/#reporting-bugs) and include the crash backtrace.
Stack dump:
0.	Program arguments: /home/butta/swift-5.8-DEVELOPMENT-SNAPSHOT-2023-01-24-a-ubuntu20.04/usr/bin/swift-frontend -frontend -c -primary-file /home/butta/swift-nio/Sources/NIOCrashTester/CrashTestSuites.swift -primary-file /home/butta/swift-nio/Sources/NIOCrashTester/CrashTests+ByteBuffer.swift -primary-file /home/butta/swift-nio/Sources/NIOCrashTester/CrashTests+EventLoop.swift -primary-file /home/butta/swift-nio/Sources/NIOCrashTester/CrashTests+HTTP.swift -primary-file /home/butta/swift-nio/Sources/NIOCrashTester/CrashTests+Strict.swift -primary-file /home/butta/swift-nio/Sources/NIOCrashTester/CrashTests+System.swift -primary-file /home/butta/swift-nio/Sources/NIOCrashTester/OutputGrepper.swift -primary-file /home/butta/swift-nio/Sources/NIOCrashTester/main.swift -supplementary-output-file-map /tmp/TemporaryDirectory.buZsff/supplementaryOutputs-1 -target x86_64-unknown-linux-gnu -disable-objc-interop -I /home/butta/swift-nio/.build/x86_64-unknown-linux-gnu/debug -enable-testing -g -module-cache-path /home/butta/swift-nio/.build/x86_64-unknown-linux-gnu/debug/ModuleCache -swift-version 5 -Onone -D SWIFT_PACKAGE -D DEBUG -new-driver-path /home/butta/swift-5.8-DEVELOPMENT-SNAPSHOT-2023-01-24-a-ubuntu20.04/usr/bin/swift-driver -entry-point-function-name NIOCrashTester_main -empty-abi-descriptor -resource-dir /home/butta/swift-5.8-DEVELOPMENT-SNAPSHOT-2023-01-24-a-ubuntu20.04/usr/lib/swift -enable-anonymous-context-mangled-names -Xcc -fmodule-map-file=/home/butta/swift-nio/.build/x86_64-unknown-linux-gnu/debug/CNIOSHA1.build/module.modulemap -Xcc -I -Xcc /home/butta/swift-nio/Sources/CNIOSHA1/include -Xcc -fmodule-map-file=/home/butta/swift-nio/.build/x86_64-unknown-linux-gnu/debug/CNIOLLHTTP.build/module.modulemap -Xcc -I -Xcc /home/butta/swift-nio/Sources/CNIOLLHTTP/include -Xcc -fmodule-map-file=/home/butta/swift-nio/.build/checkouts/swift-atomics/Sources/_AtomicsShims/include/module.modulemap -Xcc -I -Xcc /home/butta/swift-nio/.build/checkouts/swift-atomics/Sources/_AtomicsShims/include -Xcc -fmodule-map-file=/home/butta/swift-nio/.build/x86_64-unknown-linux-gnu/debug/CNIOAtomics.build/module.modulemap -Xcc -I -Xcc /home/butta/swift-nio/Sources/CNIOAtomics/include -Xcc -fmodule-map-file=/home/butta/swift-nio/Sources/CNIOWindows/include/module.modulemap -Xcc -I -Xcc /home/butta/swift-nio/Sources/CNIOWindows/include -Xcc -fmodule-map-file=/home/butta/swift-nio/.build/x86_64-unknown-linux-gnu/debug/CNIODarwin.build/module.modulemap -Xcc -I -Xcc /home/butta/swift-nio/Sources/CNIODarwin/include -Xcc -fmodule-map-file=/home/butta/swift-nio/.build/x86_64-unknown-linux-gnu/debug/CNIOLinux.build/module.modulemap -Xcc -I -Xcc /home/butta/swift-nio/Sources/CNIOLinux/include -module-name NIOCrashTester -o /home/butta/swift-nio/.build/x86_64-unknown-linux-gnu/debug/NIOCrashTester.build/CrashTestSuites.swift.o -o /home/butta/swift-nio/.build/x86_64-unknown-linux-gnu/debug/NIOCrashTester.build/CrashTests+ByteBuffer.swift.o -o /home/butta/swift-nio/.build/x86_64-unknown-linux-gnu/debug/NIOCrashTester.build/CrashTests+EventLoop.swift.o -o /home/butta/swift-nio/.build/x86_64-unknown-linux-gnu/debug/NIOCrashTester.build/CrashTests+HTTP.swift.o -o /home/butta/swift-nio/.build/x86_64-unknown-linux-gnu/debug/NIOCrashTester.build/CrashTests+Strict.swift.o -o /home/butta/swift-nio/.build/x86_64-unknown-linux-gnu/debug/NIOCrashTester.build/CrashTests+System.swift.o -o /home/butta/swift-nio/.build/x86_64-unknown-linux-gnu/debug/NIOCrashTester.build/OutputGrepper.swift.o -o /home/butta/swift-nio/.build/x86_64-unknown-linux-gnu/debug/NIOCrashTester.build/main.swift.o -index-store-path /home/butta/swift-nio/.build/x86_64-unknown-linux-gnu/debug/index/store -index-system-modules
1.	Swift version 5.8-dev (LLVM 08f7fecaf6cedb8, Swift 3e14777b17614d5)
2.	Compiling with the current language version
3.	While walking into 'EventLoopCrashTests' (at /home/butta/swift-nio/Sources/NIOCrashTester/CrashTests+EventLoop.swift:19:1)
4.	While walking into initializer for declaration 0xc387578 (at /home/butta/swift-nio/Sources/NIOCrashTester/CrashTests+EventLoop.swift:65:5)
5.	While walking into body of 'f()' (at /home/butta/swift-nio/Sources/NIOCrashTester/CrashTests+EventLoop.swift:81:13)
```

**Steps to reproduce**
`swift build`

**Expected behavior**
The build to succeed, as it was with the Jan. 20 snapshots

**Environment**
The official Swift 5.8 Jan. 24 snapshot and trunk Jan. 23 snapshot for Ubuntu 20.04 x86_64, building the latest NIO commit apple/swift-nio@45167b8

**Additional context**
This is most likely caused by #63127, @bnbarham, as it was one of only two pulls merged to the 5.8 branch in that time.

**Comment by bnbarham** [comment]
Looks like the assertion is actually invalid. We have a local variable declaration (which could be either a shorthand if let or a capture) and a reference to the function decl. Those do not have the same names. I haven't checked why we weren't hitting this before, presumably we were failing to actually get the referenced decl and thus never hitting this.

https://github.com/apple/swift/pull/63281
https://github.com/apple/swift/pull/63282

**Comment by buttaface** [comment]
Nothing specific to Android here, the repro above is for linux and I'm guessing it's broken on macOS too. The `compiler crash` label should be added and the platform label removed.

**Comment by buttaface** [comment]
I can confirm [this is fixed in the latest 5.8 Feb. 1 snapshot](https://github.com/buttaface/swift-android-sdk/actions/runs/4072855161/jobs/7016098220).

**Comment by bnbarham** [comment]
Thanks @buttaface :)!
